miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala

```diff
@@ -138,34 +137,123 @@ object NettyModelConversion {
+      val publisher = responseToPublisher[F](httpResponse)
       val response =
         new DefaultStreamedHttpResponse(
           httpVersion,
-          HttpResponseStatus.valueOf(http4sResponse.status.code),
+          HttpResponseStatus.valueOf(httpResponse.status.code),
           publisher
         )
-      http4sResponse.headers.foreach(h => response.headers().add(h.name.value, h.value))
+      httpResponse.headers.foreach(h => response.headers().add(h.name.value, h.value))
       response
     } else {
       val response = new DefaultFullHttpResponse(
         httpVersion,
-        HttpResponseStatus.valueOf(http4sResponse.status.code)
+        HttpResponseStatus.valueOf(httpResponse.status.code)
       )
-      http4sResponse.headers.foreach(h => response.headers().add(h.name.value, h.value))
+      httpResponse.headers.foreach(h => response.headers().add(h.name.value, h.value))
       if (HttpUtil.isContentLengthSet(response))
         response.headers().remove(`Content-Length`.name.toString())
       response
     }
-  }
+
+  private def toWSResponse[F[_]](
+      httpRequest: Request[F],
+      httpResponse: Response[F],
+      httpVersion: HttpVersion,
+      wsContext: WebSocketContext[F]
+  )(
+      implicit F: Effect[F],
+      ec: ExecutionContext
+  ): F[DefaultHttpResponse] =
+    if (httpRequest.headers.exists(
+          h => h.name.toString.equalsIgnoreCase("Upgrade") && h.value.equalsIgnoreCase("websocket")
+        )) {
+      val wsProtocol  = if (httpRequest.isSecure.exists(identity)) "wss" else "ws"
+      val wsUrl       = s"$wsProtocol://${httpRequest.serverAddr}${httpRequest.pathInfo}"
+      val bufferLimit = 65535 //Todo: Configurable. Probably param
+      val factory     = new WebSocketServerHandshakerFactory(wsUrl, "*", true, bufferLimit)
+      StreamSubscriber[F, WebSocketFrame].flatMap { subscriber =>
+        F.delay {
+          val processor = new Processor[WSFrame, WSFrame] {
+            def onError(t: Throwable): Unit = subscriber.onError(t)
+
+            def onComplete(): Unit = subscriber.onComplete()
+
+            def onNext(t: WSFrame): Unit = subscriber.onNext(nettyWsToHttp4s(t))
+
+            def onSubscribe(s: Subscription): Unit = subscriber.onSubscribe(s)
+
+            def subscribe(s: Subscriber[_ >: WSFrame]): Unit =
+              wsContext.webSocket.send.map(wsbitsToNetty).toUnicastPublisher().subscribe(s)
+          }
+
+          F.runAsync(Async.shift[F](ec) >> subscriber.stream.through(wsContext.webSocket.receive).compile.drain)(
+              _ => IO.unit
+            )
+            .unsafeRunSync()
+          val resp =
+            new DefaultWebSocketHttpResponse(httpVersion, HttpResponseStatus.OK, processor, factory)
+          wsContext.headers.foreach(h => resp.headers().add(h.name.toString(), h.value))
+          resp
+        }
+      }
+    } else {
+      wsContext.failureResponse.map(toNonWSResponse[F](_, httpVersion))
+    }
+
```

Comment:
These responses are nagging me. I think we can improve on these to some degree.

Reply:
This is encoded in the wsContext, for the client to have a particular response they want to send.